yourbench/pipeline/single_shot_question_generation.py

```diff
@@ -0,0 +1,419 @@
+from loguru import logger
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+from yourbench.utils.inference_engine import InferenceCall, run_inference
+from yourbench.utils.prompts import (
+    QUESTION_GENERATION_SYSTEM_PROMPT,
+    QUESTION_GENERATION_USER_PROMPT,
+)
+
+# Import the unified parsing function
+from yourbench.utils.parsing_engine import parse_qa_pairs_from_response
+
+@dataclass
+class SingleHopQuestionRow:
+    """
+    Represents a single-hop question row derived from a single chunk of text.
+
+    Attributes:
+        chunk_id: A string identifier for the chunk from which this question was generated.
+        document_id: Identifier for the parent document.
+        question: The generated question text.
+        self_answer: The LLM-produced short answer or reasoning.
+        estimated_difficulty: An integer from 1-10 indicating the estimated difficulty.
+        self_assessed_question_type: A descriptor for the type or style of question.
+        generating_model: The model used to generate this question.
+        thought_process: Free-form text describing how the question was derived.
+        raw_response: The full, unedited response from the model.
+        citations: A list of references or quotes extracted from the chunk.
+    """
+
+    chunk_id: str
+    document_id: str
+    question: str
+    self_answer: str
+    estimated_difficulty: int
+    self_assessed_question_type: str
+    generating_model: str
+    thought_process: str
+    raw_response: str
+    citations: List[str]
+
+
+@dataclass
+class ChunkSamplingConfig:
+    mode: str = "all"
+    value: float = 1.0
+    random_seed: int = 42
+
+
+@dataclass
+class SingleShotQuestionGenerationConfig:
+    run: bool = False
+    source_subset: str = ""
+    output_subset: str = ""
+    additional_instructions: str = "Generate questions to test an undergraduate student"
+    chunk_sampling: ChunkSamplingConfig = field(default_factory=ChunkSamplingConfig)
+
+
+@dataclass
+class DocumentRow:
+    document_summary: str = "No summary available."
+    document_filename: str = ""
+    document_id: str = ""
+    chunks: List[Dict[str, Any]] = field(default_factory=list)
+
+
+def run(config: Dict[str, Any]) -> None:
+    """
+    Executes the Single-Shot Question Generation stage of the pipeline.
+    """
+    stage_config = _load_stage_config(config)
+    if not stage_config.run:
+        logger.info("single_shot_question_generation stage is disabled. Skipping.")
+        return
+
+    source_dataset_name, output_dataset_name, source_subset, output_subset = _resolve_dataset_info(config)
+    dataset = _load_input_dataset(source_dataset_name, config, source_subset)
+    if dataset is None:
+        return
+
+    inference_calls, call_index_mapping = _build_inference_calls(dataset, stage_config)
```

Comment:
you should add type hinting to show more clearly what you're returning, and again careful with the "if not"